core/src/program/mod.rs

```diff
@@ -147,3 +158,26 @@ impl Kernel {
         &self.0
     }
 }
+
+impl Serializable for Kernel {
+    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
+        target.write_digests(&self.0)
+    }
+}
+
+impl Deserializable for Kernel {
+    fn read_from(bytes: &mut ByteReader) -> Result<Self, SerializationError> {
+        bytes.read_digests().map(Self)
+    }
+}
+
+// this is required by AIR as public inputs will be serialized with the proof
+impl winter_utils::Serializable for Kernel {
+    fn write_into<W: winter_utils::ByteWriter>(&self, target: &mut W) {
+        // TODO this diverges from the original implementation as it serializes the length as well.
+        // given the procedures count for kernel is not constant, this is a requirement for
+        // deserialization.
+        target.write_u64(self.0.len() as u64);
+        Digest::write_batch_into(&self.0, target)
+    }
+}
```

Comment:
Similar comment about serialization as before.

Reply:
Ack! Resolving this discussion to keep the topic centralized on the previous one